### DISCUSSION TITLE
General question: contribution opportunities

### DISCUSSION THREAD

**Discussion by VladPodilnyk** [gomlx/gomlx#107] [Q&A]
Hi 👋🏻 

First of all, I'd like say thank you for creating this project and the effort you're putting into it.
I find this project is quite interesting as I believe Golang should be a nice choice for ML projects when it comes to ML infra, performance and production implementation of models.

I've glanced through the Readme, contribution guide, discussions and TODOs to find an opportunity to contribute and help with development. Unfortunately, I haven't found a clear path / tasks to start with, which is OK as I believe the most natural way would be to play with project first and then find features/places for improvements.

I don't have a lot of exposure to ML (only some basic understanding of how some of the concepts works), but I'd like to help and I can learn on the go. I've seen that there is some work needed on the infrastructure side of things, so I was hoping to try to improve that part (although I am not sure what is the current state of things, as I believe there is a related ongoing work).

With that long passage of text, I am trying to ask if there some tasks, things, even long hanging fruits that you would like to delegate or you might need a helping hand with? 🙈

**Comment by janpfeifer** [comment]
hi @VladPodilnyk , thanks for the kind words ❤️ I couldn't agree more about Golang is a nice choice to write ML code -- at the "modelling level", not at the "matrix multiplication/number crunching" level. Also because Golang is a great language to do data pre-processing, parallelization, etc.

The whole project keeps moving, and while I try to document everything, some documentation always lag behind.

To contribute, the easiest way is always having a project you want to use, or a new ML paper you would like to implement, and attempt that. This way you learn how to use GoMLX -- I'm always happy to help/answer questions where needed -- and as you bump on GoMLX limitations, there is a good place to collaborate.

Knowing ML helps a lot. ML is not that hard to learn: the math to use it and understand what is going on is not particularly hard (compared to physics let's say), and the algorithms themselves, once the math is understood, are trivial. What there is in ML is a level of intuition built with experience: contrary to normal computer algorithms, where things either work or not, in ML things sometimes work +/- ... and one doesn't know if one did something wrong or if it's simply the nature of the data or the particular problem. And for that experience helps.

There are many small tasks ... but most of them non-ML related -- for instance writing a Nix installation script for [github.com/gomlx/gopjrt](https://github.com/gomlx/gopjrt), the underlying engine that links to the XLA C++ part. But these are not ML related, so you wouldn't learn that ...

If you want to brave a more complex ML topics, one that is on my TODO list (because it looks cool and it looks promising to me) is implementing the [MARS: Unleashing the Power of Variance Reduction for Training Large Models](https://arxiv.org/abs/2411.10438](https://arxiv.org/abs/2411.10438) optimizer. Ideally, it would involve having one test model and measure if the new optimizer really shows improvement. While this is somewhat advanced ML, there is a reference implementation in PyTorch in github, making things interest.

I'm currently working on a generative model for flowers, just to learn and implement a demo using [Flow Matching](https://github.com/gomlx/gomlx/blob/main/examples/FlowMatching/flow_matching.ipynb)

Another ML topic in my TODO list, is to implement an Reinforcement Learning (for instance AlphaZero from DeepMind, or something new) for some board game using GoMLX. Take a simple board game you enjoy, implement a small engine for it, and attempt to learn a playing model. 

I hope these ideas are helpful to inspire you. For me, what usually works best, is to have a concrete goal in mind that I want to achieve, and that uses GoMLX. And through it, improve GoMLX to enable me to achieve your goal. Maybe that works for you as well ?

cheers

**Reply by VladPodilnyk** [reply]
Thanks a lot for your reply, @janpfeifer!

>I hope these ideas are helpful to inspire you. For me, what usually works best, is to have a concrete goal in mind that I want to achieve, and that uses GoMLX. And through it, improve GoMLX to enable me to achieve your goal. Maybe that works for you as well ?

Definitely!
I think I am not comfortable enough with ML to go after `MARS` optimiser now, but I'll definitely take a look at this paper 🙌🏻 
On the other hand, I think playing with reinforcement learning seems like an awesome and fun idea for a small side-project.

Also, this reminded me, that for a long time I wanted to understand how neural style transfer works. I found these [paper](https://arxiv.org/abs/1508.06576) and [repo](https://github.com/jcjohnson/neural-style?tab=readme-ov-file) that might be useful.
So, I think I can implement such algorithm using GoMLX and create a small demo web app with it.

How do you think, will it be useful for you as well?

**Comment by janpfeifer** [comment]
hey @VladPodilnyk , apologies the the delay responding, I got caught up on some other projects and lost track.

The style transfer sounds like a great project, and super fun!

A few comments -- from just guessing:

1. The original paper is very old, and there are probably much better ways of doing it. But you can still start with it, just probably you want to use a pre-trained image model. GoMLX already includes a converted [InceptionV3, located here](https://github.com/gomlx/gomlx/tree/main/models), that you can use. 
2. For different pre-trained image models, and if you don't want to manually convert them like i did with InceptionV3, there is a ONNX to GoMLX converter in [github.com/gomlx/onnx-gomlx/](https://github.com/gomlx/onnx-gomlx/), and allows you to use any model in HuggingFace that is (or has) saved in ONNX format -- it's kind of new and we've been using mostly for LLM, let me know if you find any issues.
3. In either case (InceptionV3 or some ONNX converted model) you will need to do some model "surgery" to find the nodes that represent the output of each layer. The InceptionV3 you can change the code to optionally include the output of all layers in what it returns. For ONNX models you will probably want to do it manually, by plotting the computation graph (it includes a tool to output that), and figure out the node for each layer -- after you identify the first, the others are usually very similarly named.
4. If you get it to work, pls share, I'd love to see it in action! 😃   And I'd love to at least link your example in GoMLX, or include it under the `examples/` directory 

As an example I think it would be super useful. Plus, despite being old, it's still a fun party trick -- you can connect to a camera and do cartoonish videos. I wonder if it would work as a plugin for video chats (Google's meet or zoom)...

**Reply by VladPodilnyk** [reply]
> apologies the the delay responding, I got caught up on some other projects and lost track.

No worries at all, thanks for taking a time to answer my questions 🙌🏻
I appreciate that 🙇🏻 

> it's still a fun party trick -- you can connect to a camera and do cartoonish videos. I wonder if it would work as a plugin for video chats (Google's meet or zoom)...

It would be just AWESOME xD. I'd love to make this!

Regarding your suggestions, I have a rookie question 🙈 (currently I have a lot of entropy about the project, and I believe I'll shed some light on it, later on while implementing it.)
You mentioned that I might need to find output nodes for each layer. I am trying to guess, I might need this to attach "new" layers to the outputs that I have found, correct?

**Comment by janpfeifer** [comment]
Yes, kind of. A `*graph.Node` represents the output of a computation in the graph. In style transfer there are 3 images:

1. Style image, let's call it $i_s$
2. Original image you want to style, let's call it $i_o$
3. Image you want to generate: $x$

You pass each of these images through the same image model -- like an image classifier, let's say InceptionV3 -- notice you can take the same model, and apply it 3 times to different outputs.

Now, somewhere in the IncpetionV3 code, there is a computation with the output of the each layer -- likely in the middle of a for loop. It will be a variable of type `*Node`. I think the easiest first approach is to change the code of the InceptionV3, to also returns those `*Node` values, that represent the output of each layer. Maybe by creating an alternative InceptionV3 function that returns them.

With the output of each layer (the *Node variables)  for each of the images, you can start playing with adding the losses the paper suggest. For style, you add a L2 loss (MeanSquaredError) on the lower layers on the difference of the Layer(x) and Layer($i_s$). And for the image content so to-say, an L2 loss on the upper layer(s), but on the difference of the Lyaer(x) and Layer($i_o$).

The paper also suggests other losses. And since the paper is old, I bet there are even better ways to do style transfer, one would need to search around. But it's nice to start with the original paper.

Now with the losses defined, if I understood correctly, you want to define $x$ (the image being generated) as a variable to learn. Just use the Context to create a variable with the shape of an image, and initialize it with the original image (I think), see `Context.VariableWIthValue()`. Remember to freeze the InceptionV3 model -- u don't want to change the image model, you are just "learning" the value of $x$. I think the InceptionV3 library has an option to do that for you.

After you get the basics going -- it should do something, probably not as good as the paper initially, you can go check the other losses proposed in the paper. I'm not sure how much each loss contribute, I haven't done that before ...

**Reply by VladPodilnyk** [reply]
Thanks a lot, @janpfeifer, for a such detailed answers and for clarifying the whole picture a bit.
I definitely feel that I have gaps in my knowledge and understanding some probably some of the things went over my head 🙈  (but they started a thought process which is great).

I think I'll start with the original paper and your suggestions (check the code as well). I guess, it's a good starting point.
Hopefully, I'll come back to you as soon as I get a decent results 😌.

**Comment by janpfeifer** [comment]
Also, remember the AI is your friend 😃  ... sometimes a not very smart one (it hallucinates a lot), but it often helps.

**Reply by VladPodilnyk** [reply]
Got it, thanks 🙌🏻

**Comment by janpfeifer** [comment]
hey @VladPodilnyk , just a heads up.

To make it easy for you to capture the output of the InceptionV3 model -- or to expose middle of the model nodes yourself -- I created the concept of a "node alias". So the InceptionV3 now optionally (if you configure it with `WithAliases(true)`) will create aliases to the output of every convolution layer.

From the `CHANGELOG`:

  * Aliases nodes: allow setting aliases to nodes, and to retrieve them by those aliases. Useful for layers
    or models to export intermediary nodes by their aliases. They are prefixed by scope. New methods are:
    `Node.WithAlias`, `Node.GetAlias`, `Graph.GetNodeByAlias`, `Graph.PushAliasScope`, `Graph.PopAliasScope` 
    and `Graph.IterAliasedNodes`.
  * Added optional aliases nodes for `inceptionv3` model.

It will be part of the next release (probably next week), but it is already merged into the `main` branch of the project.

Here the PR with the changes, if you are curious: https://github.com/gomlx/gomlx/pull/110

And here the node aliases created by inceptionV3 (you can use `Graph.GetNodeByAlias` to get the corresponding node):

```
"inceptionV3/conv_000/output", "inceptionV3/conv_001/output", ..., "inceptionV3/conv_093/output",
"inceptionV3/logits"
```

**Reply by VladPodilnyk** [reply]
Wow, that's awesome 🙌🏻 
Thanks!

I've checked your PR and I've realised that inceptionV3 expects images of size 299x299, right?
Here I might have a dumb question 🙈 So, let's say in case of a plugin that cartoonify video stream, each frame is larger that the in put size of the model, which means that it should be downscaled to fit the inputs. Then if I just upscale the output image it might look very terrible due to information loss. I understand that I am getting ahead of myself, but now I am wondering do I need to add some kind of additional layers to adapt the model to images of lager sizes?

Sorry, if my questions are super rookie 🙈 

p.s. I like how well documented the library is 🚀

**Reply by janpfeifer** [reply]
> Here I might have a dumb question 🙈 So, let's say in case of a plugin that cartoonify video stream, each frame is larger that the in put size of the model, which means that it should be downscaled to fit the inputs. Then if I just upscale the output image it might look very terrible due to information loss. I understand that I am getting ahead of myself, but now I am wondering do I need to add some kind of additional layers to adapt the model to images of lager sizes?

That's actually a very good question, that the paper doesn't discuss in details. I assume they just rescaled/cropped the images to the VGG input size (224x224). And you are right, once down-scaled you loose quality.

My recommendation is just starting with rescaling to the small image to see if things work. But after you get that to work, then some suggestions from the top of my mind:

- Tile the image, and do generation on the tiles, specially for the style. 
- But for the original image, use a scale -- you only want to preserve the high level idea of the original image, so scaling may be enough. There is an `Interpolate` op in GoMLX that I think is differentiable. So it's just like a layer you put in front of the image being generated (at higher resolution), then feed the interpolated (down-scaled) result to to InceptionV3, and do the loss there. It should all back-propagate correctly to the image being generated (crossing-fingers).
- If the tiling for the style matching leads to inconsistencies at the border, you can always use overlapping tiles, then it will presumably blend somewhat.

Now ... there are newer approaches for style transfer that I think don't suffer from this issue. But I haven't read about them yet, so I don't know. Maybe an alternative next step after you get the simpler down-scaled version working would be to investigate ?  I saw the title of a paper from 2024 on the topic the other day.

**Reply by VladPodilnyk** [reply]
Thanks for your answer and suggestions 🙌🏻 

>Maybe an alternative next step after you get the simpler down-scaled version working would be to investigate ? I saw the title of a paper from 2024 on the topic the other day.

Yes, sure 😌 I just didn't think about this at first, but then while I was looking at your PR it just popped into my head.
I'll explore.

Thanks again!

**Reply by janpfeifer** [reply]
So, yesterday I played around with it a bit, after [reading this blog post](https://medium.com/@ferlatti.aldo/neural-style-transfer-nst-theory-and-implementation-c26728cf969d), to make sure you would have everything you need from GoMLX. 

My findings:

* I had to do a one-line fix in the InceptionV3 code (to make it differentiable even if its weighs are frozen), and I already submitted to head in GoMLX (but I didn't cut a new release).
* One has to normalize the image color values from -1 to 1 before feeding to InceptionV3 (usually images are loaded with values from 0 to 1). And the de-normalize the generated image.
* During the transfer learning, it used to generate image values > 1.0 or < -1.0. That lead to "specks" (pixels with too large values) in the images. I fixed it by after each step of transfer I had to clip the values back to be between -1 and 1.0.

Here a sample of the results of a hacky version, as a proof of concept:

![image](https://github.com/user-attachments/assets/355f94ac-9806-42a5-87f9-6488a8305b5a)

Probably there are tons that can be improved -- including handle the larger resolutions. I'm hoping you'll figure it out 😄

**Reply by VladPodilnyk** [reply]
Wow 😮 
That's look awesome 🙌🏻 

Thanks for sharing it and for you findings, they are very insightful! 

>Probably there are tons that can be improved -- including handle the larger resolutions. I'm hoping you'll figure it out 😄

I will!

**Reply by janpfeifer** [reply]
Another quick update: I cleaned up the code and put it out in [a small repository (github.com/janpfeifer/styletransfer)](https://github.com/janpfeifer/styletransfer). I noticed it is 10 years of the original paper, I thought it would make a nice "celebratory" demo. 

Very simple ... fixed resolution and slow ... but may help bootstrap your work/project...

**Reply by VladPodilnyk** [reply]
> but may help bootstrap your work/project...

100%, as I am currently going though papers and reading Python implementations using Pytorch. 
I am going a bit slow, but as I am trying to build an intuition about how different concepts work it's fine, I guess 🙈 
Having a reference implementation in Golang and GoMLX will help a lot! 
Thank you 🙌🏻
